### ISSUE TITLE
mkdir: can't create directory '/home/qwc': Permission denied

### ISSUE THREAD

**Issue by rsrg-zwiama** [qwc-services/qwc-uwsgi-base#4]
Due to the update in the qwc-uwsgi-base commits

https://github.com/qwc-services/qwc-map-viewer/commit/a376812c04e2bbd972313c267106ec9d8eb5583d
and 
https://github.com/qwc-services/qwc-uwsgi-base/commit/7848c2868d4dc9523f941c29eaffd61456106f6b 

Pods based on the sourcepole/qwc-uwsgi-base:alpine-v2023.06.08 are failing to start when running with a user other than 0 in AKS with the error: `mkdir: can't create directory '/home/qwc': Permission denied`
Tested with  sourcepole/qwc-map-viewer-base  v2023.08.31 and sourcepole/qwc-db-auth  v2023.06.08

We run qwc-services in Azure Kubernetes Service with
``` 
 securityContext:
    runAsUser: 1012
    runAsGroup: 1012
.
.
.
  SERVICE_UID:
    value: "1012"
  SERVICE_GID:
    value: "1012"


```
Best,
Amanda

**Comment by manisandro** [comment]
This is because of [1] specifically because `libpq` in alpine-3.18 as of July 2023 will fail with DB connections with `sslmode=required` attempting to look for certificates below a non-existing home directory. Annoyingly, it requires the home directory to exist, even if no certificates exist below it.

I could not find another solution. Perhaps it has been fixed with newer libpq...

[1] https://github.com/qwc-services/qwc-uwsgi-base/blob/main/alpine/Dockerfile#L21

**Comment by rsrg-zwiama** [comment]
Thank you for the info. For us it's not a problem to have this folder, but is there a way to change permissions fist. So the AKS User has permission to create the folder?

Thanks,
Amanda

**Comment by manisandro** [comment]
Unfortunately I don't see any way to do this. You need root to `chown`, which can only be executed at runtime because `SERVICE_UID/SERVICE_GID` are ENVs.

**Comment by rsrg-zwiama** [comment]
I see the point, so we will have to change on our side. Thanks for your help, I will inform you if we find an alternative solution.

**Comment by danduk82** [comment]
Hello @manisandro ,
what do you think about this https://github.com/qwc-services/qwc-uwsgi-base/pull/5 ?
In this PR I set `/tmp` as `HOME` for the qwc user. I don't know what side effect this can have in other services using this image, is the HOME directory ever used? If not this should work, but I didn't test it extensively.

What do you think?

**Comment by manisandro** [comment]
I seem to remember that I tried it at the time and it didn't work. I did a quick test with a current libpq dropping the home directory completely and it seemed to work, possibly libpq has since improved it's handling of a missing home directory. Possibly we can just drop the entire mkdir thing again in time for 2024-lts.

**Comment by danduk82** [comment]
ah well, this would be the better indeed

**Comment by rsrg-zwiama** [comment]
Short question to this again: The changes from https://github.com/qwc-services/qwc-uwsgi-base/pull/5 aren't in the alpine-latest docker image (and for that reason also not used by oidc-service (latest). Can you update the images, so changes get available? 
https://hub.docker.com/layers/sourcepole/qwc-uwsgi-base/alpine-latest/images/sha256-ba300f53cc53a423654f43fb21bfdd274dfe1d2e8e011692a6e289ccb8cce640?context=explore

**Comment by manisandro** [comment]
I've tagged qwc-oidc-auth-v2023.11.30 which uses a qwc-uwsgi-base-alpine with this change.

**Comment by rsrg-zwiama** [comment]
Thanks for the fast reply, but unfortunately the new build crashes with the error: 
ImportError: cannot import name 'url_quote' from 'werkzeug.urls' (/usr/lib/python3.11/site-packages/werkzeug/urls.py)

looks like this issue: https://stackoverflow.com/questions/77213053/why-did-flask-start-failing-with-importerror-cannot-import-name-url-quote-fr

**Comment by manisandro** [comment]
Right, please try out v2023.11.30.1

**Comment by rsrg-zwiama** [comment]
The services starts/runs but in the login process it fails: 
```
[2023-11-30 13:29:27,273] ERROR in app: Exception on /callback [GET]
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/srv/qwc_service/server.py", line 191, in callback
    access_token = create_access_token(identity)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/flask_jwt_extended/utils.py", line 169, in create_access_token
    return jwt_manager._encode_jwt_from_config(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/flask_jwt_extended/jwt_manager.py", line 506, in _encode_jwt_from_config
    json_encoder=config.json_encoder,
                 ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/flask_jwt_extended/config.py", line 276, in json_encoder
    return current_app.json_encoder
           ^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Flask' object has no attribute 'json_encoder'

```
I tried` werkzeug==2.2.2` what works on first sight....

**Comment by manisandro** [comment]
Should be fixed by updating to Flask-JWT-Extended==4.4.4 => v2023.11.30.2

**Comment by rsrg-zwiama** [comment]
great, all working now! 
Thanks
